### PR TITLE
[Merged by Bors] - feat(AlgebraicGeometry/ProjectiveSpectrum/Scheme): `fromSpec` and `toSpec` compose to identity

### PR DESCRIPTION
**Pull request by jjaassoonn**
---

- [x] depends on: #9601

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Comment by erdOne** [comment]
The proofs seem to be quite long. Is it possible to extract useful lemmas?

**Comment by jjaassoonn** [comment]
> The proofs seem to be quite long. Is it possible to extract useful lemmas?

The length is mainly due to writing an element in some span as a sum of fractions and clear denominators. I will think about if any of these can be extracted out.

**Comment by jjaassoonn** [comment]
> The proofs seem to be quite long. Is it possible to extract useful lemmas?

I golfed one lemma `ToSpec.carrier_eq_span` in #9601. Now #9601 is < 100 lines long which I think is an okay length. Could you please take a look at #9601?

File: Mathlib/AlgebraicGeometry/ProjectiveSpectrum/Scheme.lean [hunk 1]
```diff
@@ -662,6 +662,56 @@ lemma toSpecFromSpec {f : A} {m : ℕ} (hm : 0 < m) (f_deg : f ∈ 𝒜 m) (x :
 
 end toSpecFromSpec
 
+section fromSpecToSpec
+
+lemma fromSpecToSpec {f : A} {m : ℕ} (hm : 0 < m) (f_deg : f ∈ 𝒜 m) (x : Proj.T| pbo f) :
```
Comment (erdOne) [inline Mathlib/AlgebraicGeometry/ProjectiveSpectrum/Scheme.lean:667]:
```suggestion
lemma fromSpec_toSpec {f : A} {m : ℕ} (hm : 0 < m) (f_deg : f ∈ 𝒜 m) (x : Proj.T| pbo f) :
```

File: Mathlib/AlgebraicGeometry/ProjectiveSpectrum/Scheme.lean [hunk 2]
```diff
@@ -662,6 +662,56 @@ lemma toSpecFromSpec {f : A} {m : ℕ} (hm : 0 < m) (f_deg : f ∈ 𝒜 m) (x :
 
 end toSpecFromSpec
 
+section fromSpecToSpec
+
+lemma fromSpec_toSpec {f : A} {m : ℕ} (hm : 0 < m) (f_deg : f ∈ 𝒜 m) (x : Proj.T| pbo f) :
+    FromSpec.toFun f_deg hm (toSpec x) = x := by
+  classical
+  refine Subtype.ext <| ProjectiveSpectrum.ext _ _ <| HomogeneousIdeal.ext <| Ideal.ext fun z ↦ ?_
+  fconstructor <;> intro hz
```
Comment (riccardobrasca) [inline Mathlib/AlgebraicGeometry/ProjectiveSpectrum/Scheme.lean:671]:
```suggestion
  constructor <;> intro hz
```
It seems to work, so why not.

File: Mathlib/AlgebraicGeometry/ProjectiveSpectrum/Scheme.lean [hunk 3]
```diff
@@ -662,6 +662,56 @@ lemma toSpecFromSpec {f : A} {m : ℕ} (hm : 0 < m) (f_deg : f ∈ 𝒜 m) (x :
...
+  · rw [← DirectSum.sum_support_decompose 𝒜 z]
+    refine Ideal.sum_mem _ fun i _ ↦ ?_
+    specialize hz i
+    obtain ⟨c, N, acd, eq1⟩ := ToSpec.MemCarrier.clear_denominator x hz
```
Comment (riccardobrasca) [inline Mathlib/AlgebraicGeometry/ProjectiveSpectrum/Scheme.lean:675]:
```suggestion
    refine Ideal.sum_mem _ fun i _ ↦ ?_
    obtain ⟨c, N, acd, eq1⟩ := ToSpec.MemCarrier.clear_denominator x (hz i)
```

File: Mathlib/AlgebraicGeometry/ProjectiveSpectrum/Scheme.lean [hunk 4]
```diff
@@ -662,6 +662,56 @@ lemma toSpecFromSpec {f : A} {m : ℕ} (hm : 0 < m) (f_deg : f ∈ 𝒜 m) (x :
...
+  · intro i
+    erw [ToSpec.mem_carrier_iff, HomogeneousLocalization.val_mk'']
+    dsimp only [GradedAlgebra.proj_apply]
+    rw [show (mk (decompose 𝒜 z i ^ m) ⟨f^i, ⟨i, rfl⟩⟩: Away f) =
+      (decompose 𝒜 z i ^ m : A) • (mk 1 ⟨f^i, ⟨i, rfl⟩⟩ : Away f) by
+      · rw [smul_mk, smul_eq_mul, mul_one], Algebra.smul_def]
+    exact Ideal.mul_mem_right _ _ <|
+      Ideal.subset_span ⟨_, ⟨Ideal.pow_mem_of_mem _ (x.1.asHomogeneousIdeal.2 i hz) _ hm, rfl⟩⟩
+
+lemma toSpec_injective {f : A} {m : ℕ} (hm : 0 < m) (f_deg : f ∈ 𝒜 m):
```
Comment (riccardobrasca) [inline Mathlib/AlgebraicGeometry/ProjectiveSpectrum/Scheme.lean:698]: You can maybe inline injectivity and surjectivity in the proof of bijectivity, but as you prefer.

**Review by riccardobrasca — APPROVED** [review]
Thanks!

bors d+

File: Mathlib/AlgebraicGeometry/ProjectiveSpectrum/Scheme.lean (hunk 4 above)
Comment (jjaassoonn) [reply Mathlib/AlgebraicGeometry/ProjectiveSpectrum/Scheme.lean:698]: I think I will keep them, I remembered using them in future proofs

**Pull request closed without merging**